syntax = "proto3";

import public "proto/utility.proto";

package payload;

enum PayloadType {
    ACKNOWLEDGEMENT = 0;
    REGISTRATION = 1;
    REGISTRATION_RESPONSE = 2;
    DISCOVERY_REQUEST = 3;
    DISCOVERY_RESPONSE = 4;
    TASK_REQUEST = 5;
    TASK_RESPONSE = 6;
    MODEL_STATE_DICT_PARAMS = 7;
}

message PayloadMessage {
    string id = 1;
    string senderId = 2;
    utility.IpAddress senderIpAddress = 3;
    PayloadType payloadType = 4;

    oneof payloadObject {
        DiscoveryRequest discoveryRequest = 6;
        DiscoveryResponse discoveryResponse = 7;
        TaskRequest taskRequest = 8;
        TaskResponse taskResponse = 9;
        Acknowledgement acknowledgement = 10;
        Registration registration = 11;
        RegistrationResponse registrationResponse = 12;
        ModelStateDictParams modelStateDictParams = 13;
    }
}

message DiscoveryResponse {
    utility.IpAddress callerAddr = 1;
    utility.AddressTable availablePeers = 2;
}

message DiscoveryRequest { int32 peersRequested = 1; }

message TaskRequest {
    int32 numWorkers = 1;
    string leaderUuid = 2;
    utility.AddressTable assignedWorkers = 3;
    int32 numEpochs = 6;

    oneof training_data_source {
        string glob_pattern = 4;
        string training_data_index_filename = 5;
    }
}

message TrainingData {
    string training_data_index_filename = 1;
    int32 numEpochs = 2;
}

message ModelStateDictParams {
    bytes modelStateDict = 1;
    bool trainingIsComplete = 2;
}

message TaskResponse {
    bytes modelStateDict = 1;
    bool trainingIsComplete = 2;
}

message Acknowledgement {}

message Registration {}

message RegistrationResponse { utility.IpAddress callerAddr = 1; }
